File: Queue/queue.cc
#include <iostream>

using namespace std;

#define SIZE 10

typedef struct queue
{
	int end;
	int data[SIZE];
}queue;

void init(queue &q)
{
	q.end = 0;
}

bool isFull(queue &q)
{
	return q.end == SIZE;
}

bool isEmpty(queue &q)
{
	return q.end == 0;
}

void insert (queue &q, const int &d)
{
	if(!isFull(q))
	{
		q.data[q.end++] = d;
	}
}

void shiftQueue (queue &q)
{
	for(int i = 0; i < q.end - 1; i++)
	{
		q.data[i] = q.data[i + 1];
	}
	q.end--;
}

int remove (queue &q)
{
	if(!isEmpty(q))
	{
		int d = q.data[0];
		shiftQueue(q);
		return d;
	}
	return -1;
}

int main()
{
        queue q;
        init(q);
        int i = 0;
        while(!isFull(q))
        {
                cout <<"I am adding "<<i<<" to queue"<<endl;
                insert(q, i);
                i++;
        }
        while(!isEmpty(q))
        {
                cout<<"I have removed "<<remove(q)<<" from queue"<<endl;
        }
return 0;
}

File: Java/Queue/Queue.java

class Queue
{
	private int []list;
	private int head;
	private int tail;
	
	public Queue(final int size)
	{
		list = new int[size];
		head = -1;
		tail = -1;
	}
	
	public boolean isEmpty()
	{
		return -1 == head;
	}
	
	public boolean isFull()
	{
		if((tail == list.length - 1 && head == 0) || (tail + 1 == head && tail != -1))
			return true;
		return false;
	}
	
	public void insert(final int d)
	{
		if(!isFull())
		{
			if(isEmpty())
			{
				tail = head = 0;
			}
			else
			{
				++tail;
				if(tail == list.length)
					tail = 0;
			}
			list[tail] = d;
		}
	}
	public int remove()
	{
		int retVal = -1;
		if(!isEmpty())
		{
			retVal = list[head];
			++head;
			if(head == tail + 1)
				head = -1;
			else if(head == list.length)
				head = 0;
		}
		return retVal;
	}
}